### DISCUSSION TITLE
v3.0.0: Control barrier functions

### DISCUSSION THREAD

**Discussion by stephane-caron** [stephane-caron/pink#112] [Announcements]
This major release adds *Control Barrier Functions* developed by @domrachev03 and @simeon-ned :+1: Examples of what we can do with barriers include:

- [Arm: UR5](https://github.com/stephane-caron/pink/tree/b9be1feb825b39ac56a795daef0695f86cc948ca/examples/barriers#arm-ur5): with joints and end effector limits
- [Humanoid: G1](https://github.com/stephane-caron/pink/blob/37591fedbfa66df9c972c3e0c88815185bda13b9/examples/humanoid_g1_com.py): G1 humanoid squatting via a CoM task
- [Quadruped: Go2](https://github.com/stephane-caron/pink/tree/b9be1feb825b39ac56a795daef0695f86cc948ca/examples/barriers#go2-squat): Go 2 squatting with floating base position limits
- [Dual Arms: Yumi](https://github.com/stephane-caron/pink/tree/b9be1feb825b39ac56a795daef0695f86cc948ca/examples/barriers#yumi-end-effector-self-collision-avoidance): self-collision avoidance with spheres
- [Dual Arms: Iiwa](https://github.com/stephane-caron/pink/tree/b9be1feb825b39ac56a795daef0695f86cc948ca/examples/barriers#iiwa-whole-body-collision-avoidance): whole-body self-collision avoidance

https://github.com/domrachev03/pink/assets/28687492/d64163b6-399f-4bbf-ac50-1135fa69c2da

Another API-breaking change to the library is that *limits* are now optional and can be extended (for instance with the new acceleration limit) or disabled. New limits and tasks include:

- Limit: `AccelerationLimit`
- Task: `ComTask` for center-of-mass tracking (thanks to @simeon-ned)
- Task: `LowAccelerationTask` for smoother velocities.

Check out the full changelog below for other changes and fixes. May the Motion be with you!

### Added

- **Breaking:** Updated the logic for handling the joint limits:
  - Add a ``limits`` argument to ``build_ik`` and ``solve_ik``
  - The `check_limits` method now includes an optional `safety_break` argument to control whether execution should stop on exception.
  - The ``solve_ik`` function now includes the `safety_break` that is forwarded to `check_limits`.
- Control Barrier Functions, namely: (thanks to @domrachev03 and @simeon-ned)
  - Abstract Barrier ``Barrier``
  - Frame Position Barrier ``PositionBarrier``
  - Body Spherical Barrier ``BodySphericalBarrier``
  - Whole-body Self-Collision Avoidance Barrier ``SelfCollisionBarrier``
- Example: UR5 manipulator and GO2 quadruped robot with `PositionBarrier`
- Example: YUMI two-armed manipulator with `BodySphericalBarrier`
- Example: G1 humanoid squatting through regulating CoM.
- Limit: `AccelerationLimit`
- Task: `ComTask` for center-of-mass tracking (thanks to @simeon-ned)
- Task: `LowAccelerationTask` for smoother velocities.

### Changed

- **Breaking:** remove ``lm_damping`` parameter from DampingTask where it wasn't used
- CICD: Update ruff to 0.4.3
- Configuration accepts list of Control Barrier Functions
- DampingTask: Simplify implementation
- Update to Pinocchio 3 with 2.7 backward compatibility

### Fixed

- PostureTask: Match implementation and documentation formula
- Fix broken documentation links

### Removed

- CICD: Disable macOS checks until upstream PyPI package is fixed

<hr /><em>This discussion was created from the release <a href='https://github.com/stephane-caron/pink/releases/tag/v3.0.0'>v3.0.0</a>.</em>

**Comment by BolunDai0216** [comment]
@stephane-caron Awesome work! Wonder if you are interested in integrating CBFs for primitives more than spheres, e.g., cylinders, capsules, ellipsoids (like in [here](https://github.com/BolunDai0216/DifferentiableOptimizationCBF)).

**Reply by kassasin** [reply]
I think this example: "pink/barriers/self_collision_barrier.py" will work.We always need to get the most closest collision pair and compute jacobian.

**Comment by simeon-ned** [comment]
@BolunDai0216 If I'm not mistaken, this functionality should already work out of the box. The hpp-fcl library can parse capsules, ellipsoids, and other geometries from the Pinocchio model, as long as these geometries are included in the model. I remember we tested capsules without making specific code modifications; we simply provided an MJCF model, as Pinocchio 3 supports these. If you need, I can create a minimal example for you.

**Reply by BolunDai0216** [reply]
@simeon-ned Thanks for the quick response! If you can provide an example that would be great!
